Tribler/community/multichain/community.py

```diff
@@ -0,0 +1,251 @@
+            previous_hash_responder = self._get_latest_hash()
+
+            payload = (payload.up, payload.down, payload.total_up, payload.total_down,
+                       payload.sequence_number_requester, payload.previous_hash_requester,
+                       sequence_number_responder, previous_hash_responder)
+
+            meta = self.get_meta_message(SIGNATURE)
+
+            message = meta.impl(authentication=(message.authentication.members, message.authentication.signatures),
+                                distribution=(message.distribution.global_time,),
+                                payload=payload)
+            self.persist_signature_response(message)
+            self.chain_lock.release()
+            self._logger.info("Sending signature response.")
+            return message
+        else:
+            self._logger.info("Chain Lock: not acquired. Dropping request.")
+            return None
+
+    def allow_signature_response(self, request, response, modified):
+        """
+        We've received a signature response message after sending a request, we must return either:
+            a. True, if we accept this message
+            b. False, if not (because of inconsistencies in the payload)
+        """
+        if not response:
+            self._logger.info("Release lock: Timeout received")
+            self.chain_lock.release()
+            return False
+        else:
+            self._logger.info("Signature response received. Modified: %s" % modified)
+            return request.payload.sequence_number_requester == response.payload.sequence_number_requester and \
+                request.payload.previous_hash_requester == response.payload.previous_hash_requester and \
+                modified
+
+    def received_signature_response(self, messages):
+        """
+        We've received a valid signature response and must process this message.
+        """
+        self._logger.info("Valid %s signature responses received." % len(messages))
+        for message in messages:
+            if self.request_timeout_lock.acquire(False):
+
+                # TODO: Check expecting response
+                self.persist_signature_response(message)
+                # Release lock
+                self._logger.info("Release lock: received signature response.")
+                self.chain_lock.release()
+                self.request_timeout_lock.release()
+
+    def persist_signature_response(self, message):
+        """
+        Persist the signature response message.
+        A hash will be created from the message and this will be used as an unique identifier.
+        :param message:
+        """
+        payload = message.payload
+        (signature_requester, pk_requester) = message.authentication.signed_members[0]
+        (signature_responder, pk_responder) = message.authentication.signed_members[1]
+
+        block = DatabaseBlock([payload.previous_hash_requester, pk_requester.public_key, signature_requester,
```

Comment:
The pk must be available in Dispersy as otherwise it wouldn't have been able to verify the signature. If it accepted a message (with a signature), without having the pk available you've found a bug.